File: src/api/models/generator.py
import os
import sys
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../..")))

from typing import Any, Dict, List, Optional, Tuple, Union
from pydantic import BaseModel, Field


class BundleGeneratorRequest(BaseModel):
    offer_type: str = "atl"
    allowed_bundle_types: Tuple[str, ...] = (
        "BUNDLE_DATA",
        "BUNDLE_VOICE",
        "BUNDLE_SMS",
    )
    validity_options: Tuple[float, ...] = (24.0, 72.0, 168.0, 720.0)
    max_volume_mb: float = 25000
    max_volume_min: int = 300
    max_volume_sms: int = 150
    samples_per_type: int = Field(default=500, gt=0)
    top_n: int = Field(default=2, gt=0)


class ExistingBundleItem(BaseModel):
    bundle_id: Union[int, str]
    bundle_name: str
    price: float
    volume_mb: float
    bundle_type: str
    minutes: float
    sms: float
    avg_weekly_revenue: float
    avg_weekly_subs: float
    popularity_score: float


class GeneratedBundleItem(BaseModel):
    bundle_id: Union[int, str]
    bundle_type: str
    usage_type: Optional[str] = None
    service_class_category: Optional[str] = None
    price: float
    volume_mb: float
    volume_min: float
    volume_sms: float
    validity_hours: float
    validity_days: float
    validity_bucket: Optional[str] = None
    predicted_popularity: Optional[float] = None
    popularity_category: Optional[str] = None


class GeneratorMeta(BaseModel):
    existing_count: int
    generated_count: int
    summary_count: int
    top_per_type_count: int


class DashboardModelParams(BaseModel):
    growth_rate: float
    cannib_sensitivity: float
    price_elasticity: float
    similarity_threshold: float
    cannib_cap: float
    expected_new_market_pct: float
    cross_type_penalty: float
    feature_weights: Dict[str, float]


class BundleGeneratorResponse(BaseModel):
    request: BundleGeneratorRequest
    meta: GeneratorMeta
    model_params: DashboardModelParams
    existing_bundles: List[ExistingBundleItem]
    generated_bundles: List[GeneratedBundleItem]
    summary: List[Dict[str, Any]]
    top_per_type: List[Dict[str, Any]]
    impact_rows: List[Dict[str, Any]]

File: frontend/app/src/services/generator.ts
import axios from 'axios'

export type OfferType = 'atl' | 'diy_mode' | 'promotion_mode' | 'btl_normal' | 'btl_moderate' | 'btl_aggressive'

export type GeneratorBundleType = 'BUNDLE_DATA' | 'BUNDLE_VOICE' | 'BUNDLE_SMS'

export type BundleGeneratorRequestPayload = {
  offer_type: OfferType
  allowed_bundle_types: GeneratorBundleType[]
  validity_options: number[]
  max_volume_mb: number
  max_volume_min: number
  max_volume_sms: number
  samples_per_type: number
  top_n: number
}

export type GeneratorExistingBundleItem = {
  bundle_id: number | string
  bundle_name: string
  price: number
  volume_mb: number
  bundle_type: string
  minutes: number
  sms: number
  avg_weekly_revenue: number
  avg_weekly_subs: number
  popularity_score: number
}

export type GeneratorGeneratedBundleItem = {
  bundle_id: number | string
  bundle_type: string
  usage_type?: string | null
  service_class_category?: string | null
  price: number
  volume_mb: number
  volume_min: number
  volume_sms: number
  validity_hours: number
  validity_days: number
  validity_bucket?: string | null
  predicted_popularity?: number | null
  popularity_category?: string | null
}

export type GeneratorMeta = {
  existing_count: number
  generated_count: number
  summary_count: number
  top_per_type_count: number
}

export type GeneratorModelParams = {
  growth_rate: number
  cannib_sensitivity: number
  price_elasticity: number
  similarity_threshold: number
  cannib_cap: number
  expected_new_market_pct: number
  cross_type_penalty: number
  feature_weights: Record<string, number>
}

export type BundleGeneratorApiResponse = {
  request: BundleGeneratorRequestPayload
  meta: GeneratorMeta
  model_params: GeneratorModelParams
  existing_bundles: GeneratorExistingBundleItem[]
  generated_bundles: GeneratorGeneratedBundleItem[]
  summary: Array<Record<string, unknown>>
  top_per_type: Array<Record<string, unknown>>
  impact_rows: Array<Record<string, unknown>>
}

const API_BASE_URL =
  (import.meta.env.VITE_PRICING_API_BASE_URL as string | undefined)?.replace(/\/$/, '') ??
  'http://127.0.0.1:8000'

export const generateBundles = async (payload: BundleGeneratorRequestPayload) => {
  const response = await axios.post<BundleGeneratorApiResponse>(`${API_BASE_URL}/generator/bundles`, payload)
  return response.data
}
